from pandera import DataFrameSchema, Column, Check, Index, MultiIndex

schema = DataFrameSchema(
	columns={
		"sra-sample_name": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"^(?!\s*$).+"),
				Check.str_length(max_value=50),
			],
			nullable=False,
			unique=True,
			coerce=False,
			required=True,
			description="Identifier name used for SRA. Max length is 50 characters. Name must be unique from BioSample/Genbank submission.",
			title="sample name",
		),
		"bs-sample_name": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"^(?!\s*$).+"),
			],
			nullable=False,
			unique=True,
			coerce=False,
			required=True,
			description="Identifier name used for BioSample. Max length is 50 characters. Required if submitting to SRA.",
			title="sample_name",
		),
		"sra-file_location": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"(?i)(\W|^)(local|cloud)(\W|$)"),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Location of raw reads files. Options: \"local\" or \"cloud\".",
			title="file location",
		),
		"sra-file_1": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"^(?!\s*$).+"),
			],
			nullable=False,
			regex=True,
			unique=True,
			coerce=False,
			required=True,
			description="All file names must be unique and not contain any sensitive information. Files can be compressed using gzip or bzip2, and may be submitted in a tar archive but archiving and/or compressing your files is not required. Do not use zip! If files are stored in the cloud, provide the full cloud url. If files are stored locally, if files are stored in the folder \"<--submission_dir>/raw_reads/\", then just the file name is required; otherwise, provide the full file path for the raw read.",
			title="raw reads file name",
		),
		"sra-file_[2-9]\d*": Column(
			dtype="object",
			checks=None,
			nullable=True,
			regex=True,
			unique=False,
			coerce=False,
			required=False,
			description="Repeatable column, replace the \"#\" with the file number for each file associated with the sample. (i.e. If you had 4 separate raw reads files associated with one sample, then you would have 4 columns named: \"sra-file_1\", \"sra-file_2\", \"sra-file_3\", and \"sra-file_4\".) All file names must be unique and not contain any sensitive information. Files can be compressed using gzip or bzip2, and may be submitted in a tar archive but archiving and/or compressing your files is not required. Do not use zip! If files are stored in the cloud, provide the full cloud url. If files are stored locally, if files are stored in the folder \"<--submission_dir>/raw_reads/\", then just the file name is required; otherwise, provide the full file path for the raw read.",
			title="additional raw reads files",
		),
		"sra-library_name": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"^(?!\s*$).+"),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=False,
			description="Short unique identifier for sequencing library.",
			title="sequencing library",
		),
		"sra-loader": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"^(?!\s*$).+"),
			],
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="SRA loader selection. If NCBI says your SRA files need to be uploaded with a certain loader, you can select it here.",
			title="sra loader",
		),
		"sra-library_strategy": Column(
			dtype="object",
			checks=[
				Check.isin(["WGA","WGS","WXS","RNA-Seq","miRNA-Seq","WCS","CLONE","POOLCLONE","AMPLICON","CLONEEND","FINISHING","ChIP-Seq","MNase-Seq","DNase-Hypersensitivity","Bisulfite-Seq","Tn-Seq","EST","FL-cDNA","CTS","MRE-Seq","MeDIP-Seq","MBD-Seq","Synthetic-Long-Read","ATAC-seq","ChIA-PET","FAIRE-seq","Hi-C","ncRNA-Seq","RAD-Seq","RIP-Seq","SELEX","ssRNA-seq","Targeted-Capture","Tethered Chromatin Conformation Capture","DIP-Seq","GBS","Inverse rRNA","NOMe-Seq","Ribo-seq","VALIDATION","OTHER"]),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="The sequencing technique intended for the library.",
			title="library strategy",
		),
		"sra-library_source": Column(
			dtype="object",
			checks=[
				Check.isin(["GENOMIC","TRANSCRIPTOMIC","METAGENOMIC","METATRANSCRIPTOMIC","SYNTHETIC","VIRAL RNA","GENOMIC SINGLE CELL","TRANSCRIPTOMIC SINGLE CELL","OTHER"]),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="The type of source material that is being sequenced.",
			title="library source",
		),
		"sra-library_selection": Column(
			dtype="object",
			checks=[
				Check.isin(["RANDOM","PCR","RANDOM PCR","RT-PCR","HMPR","MF","CF-S","CF-M","CF-H","CF-T","MDA","MSLL","cDNA","ChIP","MNase","DNAse","Hybrid Selection","Reduced Representation","Restriction Digest","5-methylcytidine antibody","MBD2 protein methyl-CpG binding domain","CAGE","RACE","size fractionation","Padlock probes capture method","other","unspecified","cDNA_oligo_dT","cDNA_randomPriming","Inverse rRNA","Oligo-dT","PolyA","repeat fractionation"]),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="The method used to select and/or enrich the material being sequenced.",
			title="library selection",
		),
		"sra-library_layout": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"(?i)(\W|^)(single|paired)(\W|$)"),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Whether to expect SINGLE or PAIRED end reads. Options: \"single\" or \"paired\".",
			title="library layout",
		),
		"sra-platform": Column(
			dtype="object",
			checks=[
				Check.isin(["_LS454","ABI_SOLID","BGISEQ","CAPILLARY","COMPLETE_GENOMICS","DNBSEQ","ELEMENT","GENAPSYS","GENEMIND","HELICOS","ILLUMINA","ION_TORRENT","OXFORD_NANOPORE","PACBIO_SMRT","TAPESTRI","ULTIMA","VELA_DIAGNOSTICS"]),
			],
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Instrument platform.",
			title="platform",
		),
		"sra-instrument_model": Column(
			dtype="object",
			checks=[
				Check.isin(["454 GS","454 GS 20","454 GS FLX","454 GS FLX+","454 GS FLX Titanium","454 GS Junior","HiSeq X Five","HiSeq X Ten","Illumina Genome Analyzer","Illumina Genome Analyzer II","Illumina Genome Analyzer IIx","Illumina HiScanSQ","Illumina HiSeq 1000","Illumina HiSeq 1500","Illumina HiSeq 2000","Illumina HiSeq 2500","Illumina HiSeq 3000","Illumina HiSeq 4000","Illumina HiSeq X","Illumina MiSeq","Illumina MiniSeq","Illumina NovaSeq 6000","Illumina NovaSeq X","Illumina NovaSeq X Plus","Illumina iSeq 100","NextSeq 1000","NextSeq 2000","NextSeq 500","NextSeq 550","Complete Genomics","PacBio RS","PacBio RS II","Revio","Sequel","Sequel II","Sequel IIe","Onso","Ion Torrent PGM","Ion Torrent Proton","Ion Torrent S5 XL","Ion Torrent S5","Ion Torrent Genexus","Ion GeneStudio S5","Ion GeneStudio S5 Plus","Ion GeneStudio S5 Prime","AB 310 Genetic Analyzer","AB 3130 Genetic Analyzer","AB 3130xL Genetic Analyzer","AB 3500 Genetic Analyzer","AB 3500xL Genetic Analyzer","AB 3730 Genetic Analyzer","AB 3730xL Genetic Analyzer","GridION","MinION","PromethION","BGISEQ-50","BGISEQ-500","MGISEQ-2000RS","DNBSEQ-G400","DNBSEQ-G50","DNBSEQ-T7","DNBSEQ-G400 FAST","Element AVITI","GS111","FASTASeq 300","GenoCare 1600","GenoLab M","Tapestri","UG 100","Sentosa SQ301"]),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Instrument model used for sequencing.",
			title="instrument model",
		),
		"sra-design_description": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Brief description of materials/methods used for creating sequencing library.",
			title="design description",
		),
		"sra-title": Column(
			dtype="object",
			checks=[
				Check(lambda s: s.nunique() == 1),
			],
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Optional internal field for how the SRA submission should be named when viewed from the NCBI submission portal. If not provided, when performing submissions <--submission_name> with the suffix \"-SRA\" will be used instead.",
			title="sra submission portal name",
		),
		"sra-comment": Column(
			dtype="object",
			checks=[
				Check(lambda s: s.nunique() == 1),
			],
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Optional internal field explaining the purpose of the submission for when interacting and resolving submission issues with NCBI.",
			title="sra submission portal description",
		)
	},
     checks=None,
     index=None,
     coerce=False,
     strict="filter",
     name="sra_schema",
     ordered=False,
     unique=None,
     report_duplicates="all",
     unique_column_names=True,
     add_missing_columns=False,
     title="seqsender SRA schema",
     description="Schema validation for SRA database.",
)
