from pandera import DataFrameSchema, Column, Check, Index, MultiIndex

schema = DataFrameSchema(
	columns={
		"bs-sample_name": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"^(?!\s*$).+"),
			],
			nullable=False,
			unique=True,
			coerce=False,
			required=True,
			description="Identifier name used for BioSample. Max length is 50 characters.",
			title="sample_name",
		),
		"bs-sample_title": Column(
			dtype="object",
			checks=[
				Check.str_matches(r"^(?!\s*$).+"),
			],
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Descriptive title for sample.",
			title="sample title",
		),
		"bs-sample_description": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Optional description for sample.",
			title="sample description",
		),
		"bs-collected_by": Column(
			dtype="object",
			checks=None,
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Name of persons or institute who collected the sample",
			title="collected by",
		),
		"bs-geo_loc_name": Column(
			dtype="object",
			checks=None,
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Geographical origin of the sample; use the appropriate name from this list http://www.insdc.org/documents/country-qualifier-vocabulary. Use a colon to separate the country or ocean from more detailed information about the location, eg \"Canada: Vancouver\" or \"Germany: halfway down Zugspitze, Alps\"",
			title="geographic location",
		),
		"bs-host": Column(
			dtype="object",
			checks=None,
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="The natural (as opposed to laboratory) host to the organism from which the sample was obtained. Use the full taxonomic name, eg, \"Homo sapiens\".",
			title="host",
		),
		"bs-host_disease": Column(
			dtype="object",
			checks=None,
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Name of relevant disease, e.g. Salmonella gastroenteritis. Controlled vocabulary, http://bioportal.bioontology.org/ontologies/1009 or http://www.ncbi.nlm.nih.gov/mesh",
			title="host disease",
		),
		"bs-isolate": Column(
			dtype="object",
			checks=None,
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="identification or description of the specific individual from which this sample was obtained",
			title="isolate",
		),
		"bs-isolation_source": Column(
			dtype="object",
			checks=None,
			nullable=False,
			unique=False,
			coerce=False,
			required=True,
			description="Describes the physical, environmental and/or local geographical source of the biological sample from which the sample was derived.",
			title="isolation source",
		),
		"bs-antiviral_treatment_agent": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="What was the antiviral treatment agent?",
			title="antiviral treatment agent",
		),
		"bs-collection_device": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Instrument or container used to collect sample, e.g., swab",
			title="collection device",
		),
		"bs-collection_method": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Process used to collect the sample, e.g., bronchoalveolar lavage (BAL)",
			title="collection method",
		),
		"bs-date_of_prior_antiviral_treat": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Date of the SARS-CoV-2 antiviral treatment, e.g., 2021-03-30",
			title="date of prior SARS-CoV-2 antiviral treatment",
		),
		"bs-date_of_prior_sars_cov_2_infection": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Date of the prior SARS-CoV-2 infection, e.g., 2021-03-30",
			title="date of prior SARS-CoV-2 infection",
		),
		"bs-date_of_sars_cov_2_vaccination": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Date of the 1st dose of the SARS-CoV-2 vaccine, e.g., 2021-03-30",
			title="date of SARS-CoV-2 vaccination",
		),
		"bs-exposure_event": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Event leading to exposure, e.g., mass gathering",
			title="exposure event",
		),
		"bs-geo_loc_exposure": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The country where the host was likely exposed to the causative agent of the illness.  This location pertains to the country the host was believed to be exposed, and may not be the same as the host's country of residence, e.g., Canada",
			title="geo loc exposure",
		),
		"bs-gisaid_accession": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The GISAID accession assigned to the sequence. GISAID Accession Numbers are used as unique and permanent identifiers for each virus beginning with the letters EPI and followed by numbers, to identify viruses and/or segments; https://www.gisaid.org/; e.g., EPI_ISL_1091361",
			title="GISAID accession",
		),
		"bs-gisaid_virus_name": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The full virus name submitted to GISAID (https://www.gisaid.org/), e.g., hCoV-19/Belgium/rega-3187/2021",
			title="GISAID virus name",
		),
		"bs-host_age": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Age of host at the time of sampling",
			title="host age",
		),
		"bs-host_anatomical_material": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Host anatomical material or substance produced by the body where the sample was obtained, e.g., stool, mucus, saliva",
			title="host anatomical material",
		),
		"bs-host_anatomical_part": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Anatomical part of the host organism (e.g. tissue) that was sampled, e.g., nasopharynx",
			title="host anatomical part",
		),
		"bs-host_body_product": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="substance produced by the host, e.g. stool, mucus, where the sample was obtained from",
			title="host body product",
		),
		"bs-host_disease_outcome": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Final outcome of disease, e.g., death, chronic disease, recovery",
			title="host disease outcome",
		),
		"bs-host_health_state": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Information regarding health state of the individual sampled at the time of sampling",
			title="host health state",
		),
		"bs-host_recent_travel_loc": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The name of the country that was the destination of most recent travel. Specify the countries (and more granular locations if known) travelled, e.g., United Kingdom. Can include multiple travels; separate multiple travel events with a semicolon.",
			title="host recent travel location",
		),
		"bs-host_recent_travel_return_date": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The date of a person's most recent return to some residence from a journey originating at that residence, e.g., 2021-03-30",
			title="host recent travel return date",
		),
		"bs-host_sex": Column(
			dtype="object",
			checks=Check.str_matches(r"(?i)(\W|^)(male|female|pooled male and female|neuter|hermaphrodite|intersex|not determined|missing|not applicable|not collected|not provided|restricted access)(\W|$)"),
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Biological sex of the host",
			title="host sex",
		),
		"bs-host_specimen_voucher": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Identifier for the physical specimen. Include a URI (Uniform Resource Identifier) in the form of a URL providing a direct link to the physical host specimen. If the specimen was destroyed in the process of analysis, electronic images (e-vouchers) are an adequate substitute for a physical host voucher specimen. If a URI is not available, a museum-provided globally unique identifier (GUID) can be used. URI example:  http://portal.vertnet.org/o/fmnh/mammals?id=33e55cfe-330b-40d9-aaae-8d042cba7542; INSDC triplet example: UAM:Mamm:52179",
			title="host specimen voucher",
		),
		"bs-host_subject_id": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="a unique identifier by which each subject can be referred to, de-identified, e.g. #131",
			title="host subject id",
		),
		"bs-lat_lon": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The geographical coordinates of the location where the sample was collected. Specify as degrees latitude and longitude in format \"d[d.dddd] N|S d[dd.dddd] W|E\", eg, 38.98 N 77.11 W",
			title="latitude and longitude",
		),
		"bs-passage_method": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Description of how the organism was passaged. Provide a short description, e.g., AVL buffer+30%EtOH lysate received from Respiratory Lab. P3 passage in Vero-1 via bioreactor large-scale batch passage. P3 batch derived from the SP-2/reference lab strain. If not passaged, put \"\"not applicable\"\".",
			title="passage method",
		),
		"bs-passage_number": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The number of known passages, e.g., 3. If not passaged, put \"\"not applicable\"\".",
			title="passage number",
		),
		"bs-prior_sars_cov_2_antiviral_treat": Column(
			dtype="object",
			checks=Check.str_matches(r"(?i)(\W|^)(no|yes)(\W|$)"),
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Has the host received SARS-CoV-2 antiviral treatment?",
			title="prior SARS-CoV-2 antiviral treatment",
		),
		"bs-prior_sars_cov_2_infection": Column(
			dtype="object",
			checks=Check.str_matches(r"(?i)(\W|^)(no|yes)(\W|$)"),
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Did the host have a prior SARS-CoV-2 infection?",
			title="prior SARS-CoV-2 infection",
		),
		"bs-prior_sars_cov_2_vaccination": Column(
			dtype="object",
			checks=Check.str_matches(r"(?i)(\W|^)(no|yes - completed|yes - partially completed)(\W|$)"),
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Has the host received a SARS-CoV-2 vaccination?",
			title="prior SARS-CoV-2 vaccination",
		),
		"bs-purpose_of_sampling": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="the reason that the sample was collected, e.g., active surveillance in response to an outbreak, active surveillance not initiated by an outbreak, clinical trial, cluster investigation, environmental assessment, farm sample, field trial, for cause, industry internal investigation, market sample, passive surveillance, population based studies, research, research and development",
			title="purpose of sampling",
		),
		"bs-purpose_of_sequencing": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The reason the sample was sequenced, e.g., baseline surveillance (random sampling)",
			title="purpose of sequencing",
		),
		"bs-sars_cov_2_diag_gene_name_1": Column(
			dtype="object",
			checks=Check.str_matches(r"(?i)(\W|^)(E (orf4)|M (orf5)|N (orf9)|spike (orf2)|ORF1ab (rep)|ORF1a (pp1a)|nsp11|nsp1|nsp2|nsp3|nsp4|nsp5|nsp6|nsp7|nsp8|nsp9|nsp10|nsp12 (RdRp)|nsp13 (Hel)|nsp14 (ExoN)|nsp15|nsp16|ORF3a|ORF3b|ORF6 (ns6)|ORF7a|ORF7b (ns7b)|ORF8 (ns8)|ORF9b|ORF9c|ORF10|ORF14)(\W|$)"),
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The name of the gene used in the first diagnostic SARS-CoV-2 RT-PCR test.",
			title="SARS-CoV-2 diagnostic gene name 1",
		),
		"bs-sars_cov_2_diag_gene_name_2": Column(
			dtype="object",
			checks=Check.str_matches(r"(?i)(\W|^)(E (orf4)|M (orf5)|N (orf9)|spike (orf2)|ORF1ab (rep)|ORF1a (pp1a)|nsp11|nsp1|nsp2|nsp3|nsp4|nsp5|nsp6|nsp7|nsp8|nsp9|nsp10|nsp12 (RdRp)|nsp13 (Hel)|nsp14 (ExoN)|nsp15|nsp16|ORF3a|ORF3b|ORF6 (ns6)|ORF7a|ORF7b (ns7b)|ORF8 (ns8)|ORF9b|ORF9c|ORF10|ORF14)(\W|$)"),
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The name of the gene used in the second diagnostic SARS-CoV-2 RT-PCR test.",
			title="SARS-CoV-2 diagnostic gene name 2",
		),
		"bs-sars_cov_2_diag_pcr_ct_value_1": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The cycle threshold (CT) value result from the first diagnostic SARS-CoV-2 RT-PCR test, e.g., 21",
			title="SARS-CoV-2 diagnostic pcr Ct value 1",
		),
		"bs-sars_cov_2_diag_pcr_ct_value_2": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The cycle threshold (CT) value result from the second diagnostic SARS-CoV-2 RT-PCR test, e.g., 36",
			title="SARS-CoV-2 diagnostic pcr Ct value 2",
		),
		"bs-sequenced_by": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="The name of the agency that generated the sequence, e.g., Centers for Disease Control and Prevention",
			title="sequenced by",
		),
		"bs-vaccine_received": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Which vaccine did the host receive, e.g., Pfizer-BioNTech COVID-19 vaccine",
			title="vaccine received",
		),
		"bs-virus_isolate_of_prior_infection": Column(
			dtype="object",
			checks=None,
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Specific isolate of SARS-CoV-2 in prior infection (if known), e.g., SARS-CoV-2/human/USA/CA-CDPH-001/2020",
			title="virus isolate of prior infection",
		),
		"bs-title": Column(
			dtype="object",
			checks=[
				Check(lambda s: s.nunique() == 1),
			],
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Optional internal field for how the BioSample submission should be named when viewed from the NCBI submission portal. If not provided, when performing submissions <--submission_name> with the suffix \"-BS\" will be used instead.",
			title="biosample submission portal name",
		),
		"bs-comment": Column(
			dtype="object",
			checks=[
				Check(lambda s: s.nunique() == 1),
			],
			nullable=True,
			unique=False,
			coerce=False,
			required=False,
			description="Optional internal field explaining the purpose of the submission for when interacting and resolving submission issues with NCBI.",
			title="biosample submission portal description",
		)
	},
	checks=None,
	index=None,
	coerce=False,
	strict="filter",
	name="biosample_package_SARS-CoV-2.cl.1.0_schema",
	ordered=False,
	unique=None,
	report_duplicates="all",
	unique_column_names=True,
	add_missing_columns=False,
	title="BioSample package SARS-CoV-2.cl.1.0 schema",
	description="Schema validation for BioSample database using SARS-CoV-2.cl.1.0 package.",
)